File: src/checks/drawerpresentation/voicemenus.cpp
// Voice Changes context menu on the shared canvas popup session: the Quick
// surface that replaced the native Insert / Change voice / Delete menu.
// Every scenario drives the real right-press, the typed rows, and the shared
// picker; rows are located by typed id and clicked through the live panel,
// never the model. The edges pin the open-time capture: camera scrolls never
// drift or kill it, rewrites between a row's press and release reject the
// pick, outside right presses dismiss without the note menu's retarget, and
// a foreign takeover strands the displaced target.

#include "checks/drawerpresentation/tst_drawerpresentation.h"

#include <QCoreApplication>
#include <QEnterEvent>
#include <QPointer>
#include <QQuickItem>
#include <QQuickWindow>
#include <QtTest>

#include <cstdint>
#include <utility>

#include "checks/drawerpresentation/fixtures.h"
#include "checks/quickpopupguard.h"
#include "checks/voicepickerdriver.h"
#include "core/songdocument.h"
#include "ui/editordrawer/voicechangearea/voicechangearea.h"
#include "ui/songview.h"
#include "ui/songview/quick/timelineinputitem.h"
#include "ui/songview/quick/timelinequickview.h"
#include "ui/songview/timeruler.h"

using namespace checks::drawerpresentation;

namespace {

using VoiceMenuAction = VoiceChangeArea::VoiceMenuAction;

// The staged marker (tick 48, program 3) is the marker target; tick 144 is
// the empty-lane witness.
constexpr Tick kMarkerTick = 48;
constexpr int kMarkerVoice = 3;
constexpr Tick kEmptyTick = 144;

// One opened voice menu: session, rendered panel's typed row model, resolved
// rows, open-failure diagnostic.
struct VoiceMenu {
    songview::QuickPopupSession *session = nullptr;
    songview::QuickMenuModel *model = nullptr;
    int insertRow = -1;
    int changeRow = -1;
    int deleteRow = -1;
    QString diagnostic;
};

// Opens the voice menu through the real band right-press: live on the press
// itself, the paired release picking nothing.
VoiceMenu openVoiceMenu(VoiceTransactionFixture &fixture, Tick tick, QString diagnostic)
{
    VoiceMenu menu;
    menu.diagnostic = std::move(diagnostic);
    const QPointF point(fixture.xForTick(double(tick)), fixture.bandRect().height() / 2.0);
    sendMouse(fixture.input(), QEvent::MouseButtonPress, point, Qt::RightButton, Qt::RightButton);
    const QPointer<songview::QuickPopupSession> live{quick_popup::popupSession(fixture.view())};
    if (!QTest::qWaitFor([&live] {
            return live && live->isOpen() && quick_popup::menuPanel(*live) &&
                   quick_popup::menuModel(*quick_popup::menuPanel(*live)) != nullptr;
        }))
        return menu;
    menu.session = live;
    menu.model = quick_popup::menuModel(*quick_popup::menuPanel(*live));
    menu.insertRow = menu.model->rowForId(int(VoiceMenuAction::InsertVoiceChange));
    menu.changeRow = menu.model->rowForId(int(VoiceMenuAction::ChangeVoice));
    menu.deleteRow = menu.model->rowForId(int(VoiceMenuAction::DeleteMarker));
    sendMouse(fixture.input(), QEvent::MouseButtonRelease, point, Qt::RightButton);
    pump();
    return menu;
}

// Window-level right delivery: the open menu's frame and session see the
// press before any band does.
void windowPress(VoiceTransactionFixture &fixture, QQuickWindow &window, const QPointF &itemPoint)
{
    const QPoint point = fixture.input().mapToScene(itemPoint).toPoint();
    QEnterEvent enter(point, point, window.mapToGlobal(point));
    QCoreApplication::sendEvent(&window, &enter);
    QTest::mouseEvent(QTest::MouseMove, &window, Qt::NoButton, Qt::NoModifier, point);
    QTest::mouseEvent(QTest::MousePress, &window, Qt::RightButton, Qt::NoModifier, point);
}

void windowRelease(VoiceTransactionFixture &fixture, QQuickWindow &window, const QPointF &itemPoint)
{
    const QPoint point = fixture.input().mapToScene(itemPoint).toPoint();
    QTest::mouseEvent(QTest::MouseRelease, &window, Qt::RightButton, Qt::NoModifier, point);
}

} // namespace

void DrawerPresentationTest::voiceContextMenuTransactions()
{
    VoiceTransactionFixture fixture;
    createVoiceFixture(fixture);
    quick_popup::PromptGuard guard(fixture.view());
    SongView &view = fixture.view();
    const Snapshot before = fixture.snapshot();

    // Empty lane: the Insert pick hands the captured target to the shared picker.
    const VoiceMenu inserted =
        openVoiceMenu(fixture, kEmptyTick,
                      QStringLiteral("the empty-lane right-press did not open the voice menu"));
    QVERIFY2(inserted.session, qUtf8Printable(inserted.diagnostic));
    QVERIFY2(quick_popup::clickMenuRow(*inserted.session, inserted.insertRow),
             "the Insert row did not receive a real click");
    checks::voicepicker::Picker picker;
    QTRY_VERIFY((picker = checks::voicepicker::active(view)));
    QVERIFY2(!quick_popup::menuPanel(*picker.session),
             "the Insert pick left the voice menu panel up");
    checks::voicepicker::filter(picker, QStringLiteral("007"));
    QTRY_VERIFY(checks::voicepicker::row(picker, 7));
    checks::voicepicker::accept(picker);
    QTRY_VERIFY(!checks::voicepicker::active(view));

    DocLanePoint point;
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QCOMPARE(point.value, 7);
    QCOMPARE(fixture.document().revision(), before.revision + 1);
    QCOMPARE(fixture.document().undoStack()->index(), before.undoIndex + 1);

    // Marker lane: Change Voice reuses the direct double-click picker with the
    // captured marker voice preselected.
    const Snapshot changed = fixture.snapshot();
    const VoiceMenu changeMenu = openVoiceMenu(
        fixture, kEmptyTick, QStringLiteral("the marker right-press did not open the voice menu"));
    QVERIFY2(changeMenu.session, qUtf8Printable(changeMenu.diagnostic));
    QVERIFY2(quick_popup::clickMenuRow(*changeMenu.session, changeMenu.changeRow),
             "the Change Voice row did not receive a real click");
    QTRY_VERIFY((picker = checks::voicepicker::active(view)));
    QCOMPARE(picker.list->property("currentIndex").toInt(), 7);
    checks::voicepicker::filter(picker, QStringLiteral("003"));
    QTRY_VERIFY(checks::voicepicker::row(picker, 3));
    checks::voicepicker::accept(picker);
    QTRY_VERIFY(!checks::voicepicker::active(view));
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QCOMPARE(point.value, 3);
    QCOMPARE(fixture.document().revision(), changed.revision + 1);
    QCOMPARE(fixture.document().undoStack()->index(), changed.undoIndex + 1);

    // The Delete pick removes exactly the captured marker.
    const Snapshot changedAgain = fixture.snapshot();
    const VoiceMenu deleteMenu =
        openVoiceMenu(fixture, kEmptyTick,
                      QStringLiteral("the marker right-press did not reopen the voice menu"));
    QVERIFY2(deleteMenu.session, qUtf8Printable(deleteMenu.diagnostic));
    QVERIFY2(quick_popup::clickMenuRow(*deleteMenu.session, deleteMenu.deleteRow),
             "the Delete row did not receive a real click");
    QVERIFY2(!deleteMenu.session->isOpen(), "the Delete pick left the voice menu open");
    QVERIFY(!fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QCOMPARE(fixture.document().revision(), changedAgain.revision + 1);
    QCOMPARE(fixture.document().undoStack()->index(), changedAgain.undoIndex + 1);

    // Undo walks the three menu transactions back to the untouched song.
    fixture.document().undoStack()->undo();
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QCOMPARE(point.value, 3);
    fixture.document().undoStack()->undo();
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QCOMPARE(point.value, 7);
    fixture.document().undoStack()->undo();
    QVERIFY(!fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QVERIFY(fixture.document().smf().write() == before.smf);
}

// A camera-only scroll after the menu opens neither drifts the captured
// target nor closes the menu.
void DrawerPresentationTest::voiceMenuTargetHoldsAcrossCameraScroll()
{
    VoiceTransactionFixture fixture;
    createVoiceFixture(fixture);
    quick_popup::PromptGuard guard(fixture.view());
    SongView &view = fixture.view();
    const Snapshot before = fixture.snapshot();
    // Full lane baseline: the SMF already stages program state, so never assume a count.
    const auto beforePoints = fixture.document().lanePoints(0, DOC_CC_VOICE);

    const VoiceMenu menu = openVoiceMenu(
        fixture, kMarkerTick, QStringLiteral("the marker right-press did not open the voice menu"));
    QVERIFY2(menu.session, qUtf8Printable(menu.diagnostic));

    view.setEditorHorizontalScroll(view.camera().scrollX() + 200.0);
    pump();
    QVERIFY2(menu.session->isOpen(), "a camera-only scroll closed the useful voice menu");
    QVERIFY2(quick_popup::clickMenuRow(*menu.session, menu.changeRow),
             "the scrolled Change Voice row did not receive a real click");
    checks::voicepicker::Picker picker;
    QTRY_VERIFY((picker = checks::voicepicker::active(view)));
    QVERIFY2(!quick_popup::menuPanel(*picker.session),
             "the Change pick left the voice menu panel up");
    QCOMPARE(picker.list->property("currentIndex").toInt(), kMarkerVoice);
    checks::voicepicker::filter(picker, QStringLiteral("007"));
    QTRY_VERIFY(checks::voicepicker::row(picker, 7));
    checks::voicepicker::accept(picker);
    QTRY_VERIFY(!checks::voicepicker::active(view));

    // Exactly the captured marker moved to 7; every pre-existing unrelated
    // point stands untouched.
    const auto afterPoints = fixture.document().lanePoints(0, DOC_CC_VOICE);
    QCOMPARE(afterPoints.size(), beforePoints.size());
    for (const DocLanePoint &beforePoint : beforePoints) {
        DocLanePoint afterPoint;
        QVERIFY2(fixture.document().findLanePoint(0, DOC_CC_VOICE, beforePoint.tick, &afterPoint),
                 "a pre-existing lane point went missing across the camera scroll");
        if (beforePoint.tick == kMarkerTick)
            QCOMPARE(afterPoint.value, 7);
        else
            QCOMPARE(afterPoint.value, beforePoint.value);
    }
    QCOMPARE(fixture.document().revision(), before.revision + 1);
    QCOMPARE(fixture.document().undoStack()->index(), before.undoIndex + 1);
}

// A rewrite between a row's press and its release makes the activation
// stale: exactly the rewrite stands.
void DrawerPresentationTest::voiceMenuStaleDocumentRejectsPick()
{
    VoiceTransactionFixture fixture;
    createVoiceFixture(fixture);
    quick_popup::PromptGuard guard(fixture.view());
    SongView &view = fixture.view();

    const VoiceMenu menu = openVoiceMenu(
        fixture, kMarkerTick, QStringLiteral("the marker right-press did not open the voice menu"));
    QVERIFY2(menu.session, qUtf8Printable(menu.diagnostic));

    // The rewrite must land between the row press and its release.
    QQuickItem *const panel = quick_popup::menuPanel(*menu.session);
    QVERIFY2(panel, "the voice menu lost its rendered panel before the stale probe");
    const QPointF deleteCenter = quick_popup::menuRowSceneCenter(*panel, menu.deleteRow);
    QVERIFY2(!deleteCenter.isNull(), "the Delete row never rendered");
    QQuickWindow *const menuWindow = menu.session->window();
    QVERIFY(menuWindow);
    // Real pointer path onto the rendered delegate: enter the menu window and
    // hover the Delete row first (the same enter/move/press order the
    // windowPress helper and repo input fixtures use), so the held press
    // starts on the rendered row and the post-rewrite release completes on it.
    const QPoint deletePoint = deleteCenter.toPoint();
    QEnterEvent enter(deletePoint, deletePoint, menuWindow->mapToGlobal(deletePoint));
    QCoreApplication::sendEvent(menuWindow, &enter);
    QTest::mouseMove(menuWindow, deletePoint);
    QTest::mousePress(menuWindow, Qt::LeftButton, Qt::NoModifier, deletePoint);
    pump();
    fixture.document().addLanePoint(0, DOC_CC_VOICE, kEmptyTick, 5);
    pump();
    const Snapshot rewritten = fixture.snapshot();

    QTest::mouseRelease(menuWindow, Qt::LeftButton, Qt::NoModifier, deletePoint);
    pump();

    // Whether the rewrite cancelled the menu or the release ran stale, the
    // outcome is identical.
    QVERIFY2(!menu.session->isOpen(), "the stale request left a popup open");
    QVERIFY2(!quick_popup::popupSession(view)->isOpen(), "the stale request reopened a popup");
    QVERIFY(fixture.snapshot() == rewritten);
    DocLanePoint point;
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kMarkerTick, &point));
    QCOMPARE(point.value, kMarkerVoice);
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kEmptyTick, &point));
    QCOMPARE(point.value, 5);
}

// The native voice menu had no outside-right retarget handler; dismiss plus
// the swallowed paired release is preserved.
void DrawerPresentationTest::voiceMenuOutsideRightDismissesWithoutRetarget()
{
    VoiceTransactionFixture fixture;
    createVoiceFixture(fixture);
    quick_popup::PromptGuard guard(fixture.view());
    SongView &view = fixture.view();
    const Snapshot before = fixture.snapshot();

    // The dismissal restores the pre-menu active-focus item: stage real
    // voice-band focus first. The fixture only stages page visibility, and
    // pointer presses never drive band focus, so without this the session
    // captures the root item and restores that.
    QVERIFY(view.focusTimelineBand(songview::TimelineBand::VoiceChanges, Qt::OtherFocusReason));
    QTRY_VERIFY(view.focusedTimelineBand() == songview::TimelineBand::VoiceChanges);
    const VoiceMenu menu = openVoiceMenu(
        fixture, kMarkerTick, QStringLiteral("the marker right-press did not open the voice menu"));
    QVERIFY2(menu.session, qUtf8Printable(menu.diagnostic));

    const QPointF miss(fixture.xForTick(double(kEmptyTick)), fixture.bandRect().height() / 2.0);
    QQuickWindow *const menuWindow = menu.session->window();
    QVERIFY(menuWindow);
    windowPress(fixture, *menuWindow, miss);
    pump();
    QVERIFY2(!menu.session->isOpen(), "an outside right press did not dismiss the voice menu");
    windowRelease(fixture, *menuWindow, miss);
    pump();
    QVERIFY2(!quick_popup::popupSession(view)->isOpen(),
             "the dismissed voice menu leaked its paired release into a new popup");
    QVERIFY(!checks::voicepicker::active(view));
    QVERIFY(fixture.snapshot() == before);
    QTRY_VERIFY2(view.quickView()->focusedBand() == songview::TimelineBand::VoiceChanges,
                 "the dismissal did not restore the voice band focus");
}

// A foreign session takeover strands the pending voice target: the ruler's
// real open method publishes live rows and the displaced target never fires.
void DrawerPresentationTest::voiceMenuForeignTakeoverStaysUsable()
{
    VoiceTransactionFixture fixture;
    createVoiceFixture(fixture);
    quick_popup::PromptGuard guard(fixture.view());
    SongView &view = fixture.view();
    const Snapshot before = fixture.snapshot();

    const VoiceMenu menu = openVoiceMenu(
        fixture, kMarkerTick, QStringLiteral("the marker right-press did not open the voice menu"));
    QVERIFY2(menu.session, qUtf8Printable(menu.diagnostic));

    QQuickItem *const root = view.quickView()->rootObject();
    QVERIFY(root);
    auto *const rulerInput =
        root->findChild<songview::TimelineInputItem *>(QStringLiteral("timelineRulerInput"));
    QVERIFY(rulerInput);
    auto *const ruler = dynamic_cast<songview::TimeRuler *>(rulerInput->interaction());
    QVERIFY(ruler);
    QQuickItem *const division =
        root->findChild<QQuickItem *>(QStringLiteral("timelineRulerDivisionControl"));
    QVERIFY(division);
    ruler->openDivisionMenu(
        division->mapToScene(QPointF(division->width() / 2.0, division->height() / 2.0)));

    const QPointer<songview::QuickPopupSession> live{quick_popup::popupSession(view)};
    QVERIFY(live);
    QVERIFY2(QTest::qWaitFor([&live] {
                 QQuickItem *const foreignPanel = quick_popup::menuPanel(*live);
                 return live->isOpen() && foreignPanel &&
                        quick_popup::menuModel(*foreignPanel) != nullptr;
             }),
             "the foreign ruler menu did not publish over the pending voice menu");

    // The foreign menu keeps working: a real row pick changes the grid.
    const int currentDenom = view.viewState().gridMinDenom;
    const int targetDenom = currentDenom == 8 ? 16 : 8;
    const int targetRow =
        quick_popup::menuModel(*quick_popup::menuPanel(*live))->rowForId(targetDenom);
    QVERIFY2(targetRow >= 0, "the foreign division menu omitted the chosen denominator");
    QVERIFY2(quick_popup::clickMenuRow(*live, targetRow),
             "the foreign division row did not receive a real click");
    pump();
    QVERIFY2(!live->isOpen(), "the foreign division pick left the shared menu open");
    QCOMPARE(view.viewState().gridMinDenom, targetDenom);
    QTRY_VERIFY2(division->hasActiveFocus(),
                 "the foreign pick did not return focus to the ruler control");

    // The displaced voice target never fired: exactly nothing was written,
    // the marker stays, and no picker surfaced late.
    QVERIFY(fixture.snapshot() == before);
    DocLanePoint point;
    QVERIFY(fixture.document().findLanePoint(0, DOC_CC_VOICE, kMarkerTick, &point));
    QCOMPARE(point.value, kMarkerVoice);
    QVERIFY(!checks::voicepicker::active(view));
    QVERIFY(!quick_popup::popupSession(view)->isOpen());
}

File: src/checks/drawerpresentation/voicemenus.swift
import Foundation
@testable import PorydawApp
import PorydawCore

// Existing scenarios paired with voicemenus.cpp.
// Entry order remains in VoiceChangesPageChecks.swift.

@MainActor
func drawerVoiceContextMenuTransactions(_ report: CheckReport, suite: DocumentSession,
                                     service: ProjectService, programs: [Int]) {
    let fixture = drawerVoiceVoiceChangesFixture(suite: suite, service: service, programs: programs)
    let page = fixture.page
    let target = VoiceOccurrence(fixture.lanePoints()[1])

    _ = page.pointerPress(x: fixture.markerX(48), y: 10, surface: 1, button: 2, modifiers: 0)
    report.expect(page.hasMenu && page.menuOpen, cppID: drawerVoiceMenuID,
                  message: "a right press on a marker opens the context menu")
    report.expectEqual([VoiceChangesPagePolicy.changeVoiceAction,
                        VoiceChangesPagePolicy.deleteMarkerAction],
                       page.menuRowActions, cppID: drawerVoiceMenuID,
                       what: "a marker target publishes the change and delete rows")
    report.expectEqual(target.text, page.menuTargetIdentity, cppID: drawerVoiceMenuID,
                       what: "the menu captured the pressed occurrence")
    report.expectEqual(48, page.menuTargetTick, cppID: drawerVoiceMenuID,
                       what: "the menu captured the marker's own tick")
    report.expect(page.interactionActive, cppID: drawerVoiceMenuID,
                  message: "an open menu reports an active interaction")

    // A camera scroll after the open neither drifts the capture nor closes it.
    fixture.session.mutateCamera { $0.setHScroll($0.snapshot.scrollX + 80) }
    report.expect(page.hasMenu, cppID: drawerVoiceMenuID, message: "a camera scroll keeps the menu open")
    report.expectEqual(target.text, page.menuTargetIdentity, cppID: drawerVoiceMenuID,
                       what: "a camera scroll does not drift the captured identity")
    report.expectEqual(48, page.menuTargetTick, cppID: drawerVoiceMenuID,
                       what: "a camera scroll does not drift the captured tick")

    // Outside dismissal writes nothing.
    let baseline = fixture.snapshot
    page.dismissVoiceMenu()
    report.expect(!page.hasMenu && !page.menuOpen, cppID: drawerVoiceMenuID,
                  message: "the outside dismissal closes the menu")
    report.expect(!page.interactionActive, cppID: drawerVoiceMenuID,
                  message: "the dismissal releases the follow-scroll gate")
    report.expectEqual(baseline, fixture.snapshot, cppID: drawerVoiceMenuID,
                       what: "the dismissal writes nothing")

    // The delete row removes exactly the captured occurrence.
    _ = page.pointerPress(x: fixture.markerX(48), y: 10, surface: 1, button: 2, modifiers: 0)
    report.expect(page.activateMenuAction(actionId: VoiceChangesPagePolicy.deleteMarkerAction),
                  cppID: drawerVoiceMenuID, message: "the delete row deletes the captured marker")
    report.expect(VoiceLanePolicy.occurrence(at: 48, in: fixture.lanePoints()) == nil,
                  cppID: drawerVoiceMenuID, message: "the deleted tick no longer holds a change")
    report.expectEqual(baseline.revision + 1, fixture.snapshot.revision, cppID: drawerVoiceMenuID,
                       what: "the deletion is one revision")
    report.expect(fixture.snapshot.canUndo && !baseline.canUndo, cppID: drawerVoiceMenuID,
                  message: "the deletion records one history entry")
    report.expectEqual([0, 120], page.markerTicks, cppID: drawerVoiceMenuID,
                       what: "the projection drops exactly the deleted marker")

    // An empty-lane target offers the insertion row, which hands the same
    // captured target to the picker.
    _ = page.pointerPress(x: fixture.markerX(96), y: 10, surface: 1, button: 2, modifiers: 0)
    report.expectEqual([VoiceChangesPagePolicy.insertVoiceChangeAction], page.menuRowActions,
                       cppID: drawerVoiceMenuID, what: "an empty-lane target publishes the insert row")
    report.expect(page.menuTargetIdentity == nil, cppID: drawerVoiceMenuID,
                  message: "the empty-lane target carries no occurrence")
    report.expect(page.activateMenuAction(actionId: VoiceChangesPagePolicy.insertVoiceChangeAction),
                  cppID: drawerVoiceMenuID, message: "the insert row opens the picker")
    report.expect(page.hasPicker, cppID: drawerVoiceMenuID, message: "the picker opens on the capture")
    report.expectEqual(96, page.pickerTargetTick, cppID: drawerVoiceMenuID,
                       what: "the picker inherits the menu's captured tick")
    report.expectEqual("Insert voice change", page.pickerTitle, cppID: drawerVoiceMenuID,
                       what: "the inherited empty-lane capture keeps the insertion title")
    report.expect(!page.hasMenu, cppID: drawerVoiceMenuID, message: "the activation consumed the menu")
    page.cancelPicker()

    // A stale menu — a rewrite between the open and the activation — writes
    // nothing, and its rows never fire.
    _ = page.pointerPress(x: fixture.markerX(120), y: 10, surface: 1, button: 2, modifiers: 0)
    report.expect(page.hasMenu, cppID: drawerVoiceMenuID, message: "the menu reopened on the third marker")
    fixture.document.writeLane(track: 0, lane: .voice, from: 0, through: 0,
                               points: [LaneWrite(tick: 0, value: programs[2])])
    page.refreshFromDocument()
    report.expect(!page.hasMenu, cppID: drawerVoiceMenuID,
                  message: "a document change cancels the open menu")
    let rewritten = fixture.snapshot
    report.expect(!page.activateMenuAction(actionId: VoiceChangesPagePolicy.deleteMarkerAction),
                  cppID: drawerVoiceMenuID, message: "an activation after the cancellation writes nothing")
    report.expectEqual(rewritten, fixture.snapshot, cppID: drawerVoiceMenuID,
                       what: "the stale activation leaves the rewrite as the only change")
}

@MainActor
func drawerVoiceOriginalMenuTransactions(_ report: CheckReport, suite: DocumentSession,
                                                service: ProjectService) {
    // [f3069ef:drawerpresentation/fixtures.cpp:44-62,406] Original MIDI fixture,
    // including the undoable setup insertion of program3 at48 before snapshot.
    let document = SongDocument(file: MidiFile(division: 24, chunks: [
        MidiChunk(events: [.meta(type: 0x51, data: [0x07, 0xA1, 0x20])], endTick: 384),
        MidiChunk(events: [.channel(status: 0xC0, data0: 0),
                           .channel(status: 0x90, data0: 60, data1: 100),
                           .channel(tick: 48, status: 0x80, data0: 60)], endTick: 384),
        MidiChunk(events: [.channel(status: 0xC1, data0: 5),
                           .channel(status: 0x91, data0: 48, data1: 100),
                           .channel(tick: 48, status: 0x81, data0: 48)], endTick: 384),
    ]), config: suite.document.state.config, source: suite.document.source,
       trackBudget: suite.document.trackBudget)
    document.writeLane(track: 0, lane: .voice, from: 48, through: 48,
                       points: [LaneWrite(tick: 48, value: 3)])
    let session = DocumentSession(document: document, service: service, lease: suite.bankLease,
                                  slots: suite.bankSlots, dirty: false,
                                  loadName: suite.bankLoadName, sampleRate: 48_000)
    session.selectedTrack = 0
    let page = VoiceChangesPage(baseFontPx: 13)
    page.attach(session: session, palette: GridPalette())
    page.configureBody(width: 1000, height: 160, gutter: 56, devicePixelRatio: 1,
                       baseFontPx: 13, dragDistance: 10)
    session.onChange = { [weak page] change in
        if !change.domains.intersection([.document, .selection, .bank]).isEmpty {
            page?.refreshFromDocument()
        }
    }
    func expect(_ condition: @autoclosure () -> Bool, _ line: Int) {
        report.expect(condition(), cppID: "drawerpresentation/DrawerPresentationTest::voiceContextMenuTransactions",
                      message: "voicemenus.cpp:\(line)")
    }
    func bytes() -> [UInt8] { try! document.captureSave().bytes }
    func value() -> Int? {
        document.lanePoints(track: 0, lane: .voice).first { $0.tick == 144 }?.value
    }
    func openMenu() {
        _ = page.pointerPress(x: session.camera.displayX(tick: 144, origin: 0, dpr: 1),
                              y: 10, surface: 1, button: 2, modifiers: 0)
    }
    let before = drawerVoiceVoiceDocumentSnapshot(document)
    let beforeBytes = bytes()
    openMenu()
    expect(page.menuOpen && page.menuTargetTick == 144, 111)
    _ = page.activateMenuAction(actionId: VoiceChangesPagePolicy.insertVoiceChangeAction)
    expect(page.pickerOpen && page.pickerTargetTick == 144, 115)
    expect(!page.menuOpen, 116)
    page.setPickerFilter(text: "007")
    expect(page.pickerRowPrograms == [7] && page.pickerIndex == 0, 119)
    _ = page.acceptPicker()
    expect(!page.pickerOpen, 121)
    expect(value() != nil, 124)
    expect(value() == 7, 125)
    expect(document.revision == before.revision + 1, 126)
    let inserted = drawerVoiceVoiceDocumentSnapshot(document)
    let insertedBytes = bytes()

    openMenu()
    expect(page.menuOpen && page.menuTargetTick == 144, 134)
    _ = page.activateMenuAction(actionId: VoiceChangesPagePolicy.changeVoiceAction)
    expect(page.pickerOpen, 137)
    expect(page.pickerIndex == 7, 138)
    page.setPickerFilter(text: "003")
    expect(page.pickerRowPrograms == [3] && page.pickerIndex == 0, 140)
    _ = page.acceptPicker()
    expect(!page.pickerOpen, 142)
    expect(value() != nil, 143)
    expect(value() == 3, 144)
    expect(document.revision == inserted.revision + 1, 145)
    let changed = drawerVoiceVoiceDocumentSnapshot(document)
    let changedBytes = bytes()

    openMenu()
    expect(page.menuOpen && page.menuTargetTick == 144, 153)
    _ = page.activateMenuAction(actionId: VoiceChangesPagePolicy.deleteMarkerAction)
    expect(!page.menuOpen, 156)
    expect(value() == nil, 157)
    expect(document.revision == changed.revision + 1, 158)

    // The original's three undo steps also prove each exact history delta:
    // every step must return the full serialized state AND checkpoint identity.
    _ = document.history.undoDocument()
    expect(bytes() == changedBytes && document.history.currentIdentity == changed.identity, 159)
    expect(value() != nil, 163)
    expect(value() == 3, 164)
    _ = document.history.undoDocument()
    expect(bytes() == insertedBytes && document.history.currentIdentity == inserted.identity, 146)
    expect(value() != nil, 166)
    expect(value() == 7, 167)
    _ = document.history.undoDocument()
    expect(bytes() == beforeBytes && document.history.currentIdentity == before.identity
        && document.history.canUndo == before.canUndo, 127)
    expect(value() == nil, 169)
    expect(bytes() == beforeBytes, 170)
}
